docs/clients/dotnet/21.2/migration-to-gRPC.md

```diff
@@ -0,0 +1,599 @@
+
+A transaction can be long-lived, and opening it for a stream doesn't lock it. Another process can write to the same stream. In this case, your transaction might fail if you use idempotent writes with the expected version. If you use transactions, we recommend reevaluating your consistency guarantees and stream modelling to reduce the need for appending events. If you still need to use them, you may consider adding your own Unit of Work implementation, as, for example:
+
+```csharp
+public class EventStoreDBUnitOfWork
+{
+    private readonly EventStoreClient eventStore;
+    private readonly List<EventData> uncommittedEvents = new();
+    private readonly string streamName;
+    private readonly StreamRevision expectedStreamRevision;
+
+    private EventStoreDBUnitOfWork(EventStoreClient eventStore, string streamName,
+        StreamRevision expectedStreamRevision)
+    {
+        this.eventStore = eventStore;
+        this.streamName = streamName;
+        this.expectedStreamRevision = expectedStreamRevision;
+    }
+
+    public static EventStoreDBUnitOfWork Begin(EventStoreClient eventStore, string streamName,
+        StreamRevision expectedStreamRevision) =>
+        new(eventStore, streamName, expectedStreamRevision);
+
+    public Task<IWriteResult> Commit(CancellationToken cancellationToken = default)
+        => eventStore.AppendToStreamAsync(
+            streamName,
+            expectedStreamRevision,
+            uncommittedEvents.ToArray(),
+            cancellationToken: cancellationToken
+        );
+
+    public void Append(params EventData[] eventData) =>
+        uncommittedEvents.AddRange(eventData);
+}
+```
+
+
+## Reading Events
+
+### Read direction
+
+TCP Client have dedicated methods for reading events in the specific direction: 
+- `ReadStreamEventsForwardAsync`,
+- `ReadStreamEventsBackwardAsync`,
+- `ReadAllEventsForwardAsync`,
+- `ReadStreamEventsBackwardAsync`.
+
+ In the gRPC client, we unified those methods into methods with the direction parameter:
+- `ReadStreamAsync`,
+- `ReadAllAsync`.
+
+To read stream forwards, use:
+
+```csharp
+await using var readResult = client.ReadStreamAsync(
+    Direction.Forwards,
+    streamName,
+    StreamPosition.Start
+);
+```
+
+To read stream backwards, use: 
+
+```csharp
+await using var readResult = client.ReadStreamAsync(
+    Direction.Backwards,
+    streamName,
+    StreamPosition.End
+);
+```
+
+Accordingly, to read the `$all` stream forwards, use:
+
+```csharp
+await using var readResult = client.ReadAllAsync(
+    Direction.Forwards,
+    streamName,
+    Position.Start
+);
+```
+
+and to  the `$all` stream backwards:
+
+```csharp
+await using var readResult = client.ReadAllAsync(
```

Comment:
Don't we want to add the max count to all of those? It will timeout if you read without max count.